File: packages/semantic-cache/src/adapters/langgraph.ts
/**
 * LangGraph BaseStore adapter for @betterdb/semantic-cache.
 *
 * BetterDBSemanticStore implements the LangGraph BaseStore interface,
 * enabling similarity-based memory retrieval from a SemanticCache instance.
 *
 * When to use this vs @betterdb/agent-cache/langgraph:
 * - Use @betterdb/agent-cache/langgraph (BetterDBSaver) for exact-match checkpoint
 *   persistence - storing and retrieving agent state snapshots by checkpoint ID.
 * - Use BetterDBSemanticStore (this class) for similarity-based memory retrieval -
 *   finding the most semantically relevant past observations, documents, or facts
 *   for a given query.
 * They can coexist on the same Valkey instance with different key prefixes.
 *
 * Storage layout:
 *   {name}:entry:{uuid}  - HSET entry per item (via SemanticCache.store)
 *   namespace is stored as the 'category' tag on entries for filtered recall
 *
 * Limitations:
 * - get() uses a Valkey SCAN for the deterministic key and may be slow on large stores.
 *   For high-frequency get() patterns, prefer a dedicated hash store (agent-cache session tier).
 * - delete() uses invalidate() which is limited to 1000 entries per call and requires
 *   FT.SEARCH to find matching keys.
 */

import type { SemanticCache } from '../SemanticCache';
import { escapeTag } from '../utils';

// --- Minimal LangGraph BaseStore shims ---
// We define minimal interface shims that match the LangGraph BaseStore contract
// without importing @langchain/langgraph-checkpoint directly, so this adapter
// works even if that package is not installed.

/** A stored item in the semantic memory store. */
export interface Item {
  /** Dot-separated namespace path (e.g. ['user', 'alice', 'memories']). */
  namespace: string[];
  /** Unique key within the namespace. */
  key: string;
  /** The stored value. */
  value: Record<string, unknown>;
  /** Creation timestamp as ISO string. */
  createdAt: string;
  /** Last update timestamp as ISO string. */
  updatedAt: string;
}

/** Options for the search() method. */
export interface SearchOptions {
  /** Natural-language query for similarity search. */
  query?: string;
  /** Maximum number of results. Default: 10. */
  limit?: number;
  /** Similarity threshold override (cosine distance, 0-2). */
  threshold?: number;
}

export interface BetterDBSemanticStoreOptions {
  /**
   * A pre-configured SemanticCache instance.
   * The cache must be initialized before calling store methods.
   */
  cache: SemanticCache;
  /**
   * Field to embed from stored values when no explicit query is provided.
   * Default: 'content'. The value of this field (if a string) is used as
   * the embedding text when put() is called without an explicit embed field.
   */
  embedField?: string;
}

function namespaceKey(namespace: string[]): string {
  return namespace.join(':');
}

function namespaceToCategory(namespace: string[]): string {
  // Replace path separators (. and /) only, leaving : intact as the namespace-segment
  // separator. Matches the Python implementation for cross-language compatibility.
  return namespaceKey(namespace).replace(/[./]/g, '_');
}

/**
 * LangGraph-compatible semantic memory store backed by SemanticCache.
 *
 * Implements a subset of the LangGraph BaseStore interface sufficient for
 * use as an in-memory store with similarity recall.
 */
export class BetterDBSemanticStore {
  private readonly cache: SemanticCache;
  private readonly embedField: string;

  constructor(opts: BetterDBSemanticStoreOptions) {
    this.cache = opts.cache;
    this.embedField = opts.embedField ?? 'content';
  }

  /**
   * Store a value at namespace/key.
   * The embedField value (if present and a string) is used as the embedding text.
   * Falls back to JSON stringified value if embedField is absent.
   */
  async put(namespace: string[], key: string, value: Record<string, unknown>): Promise<void> {
    // Upsert: remove any existing entry for this key before writing so repeated
    // put() calls don't accumulate stale duplicates.
    await this.delete(namespace, key);

    const embedText =
      typeof value[this.embedField] === 'string'
        ? (value[this.embedField] as string)
        : JSON.stringify(value);

    const category = namespaceToCategory(namespace);
    const now = new Date().toISOString();
    const item: Item = {
      namespace,
      key,
      value,
      createdAt: now,
      updatedAt: now,
    };

    await this.cache.store(embedText, JSON.stringify(item), {
      category,
      metadata: { key, namespace: namespaceKey(namespace) },
    });
  }

  /**
   * Retrieve a value by exact namespace and key.
   * Paginates through all entries in the namespace using stable SORTBY ordering.
   */
  async get(namespace: string[], key: string): Promise<Item | null> {
    const category = namespaceToCategory(namespace);
    const catFilter = `@category:{${escapeTag(category)}}`;
    const { parseFtSearchResponse } = await import('../utils');
    const BATCH = 100;
    let offset = 0;

    while (true) {
      let raw: unknown;
      try {
        raw = await this.cache._searchEntries(catFilter, BATCH, offset);
      } catch {
        return null;
      }
      const entries = parseFtSearchResponse(raw);
      if (entries.length === 0) break;

      for (const entry of entries) {
        const responseStr = entry.fields['response'];
        if (!responseStr) continue;
        try {
          const item = JSON.parse(responseStr) as Item;
          if (item.key === key) return item;
        } catch { /* skip corrupt */ }
      }

      if (entries.length < BATCH) break;
      offset += BATCH;
    }
    return null;
  }

  /**
   * Search the namespace using similarity.
   * When query is provided, embeds it and returns the k most similar entries.
   * When query is absent, returns all entries in the namespace (up to limit).
   */
  async search(namespace: string[], options?: SearchOptions): Promise<Item[]> {
    const limit = options?.limit ?? 10;
    const query = options?.query;
    const category = namespaceToCategory(namespace);

    if (query) {
      const { encodeFloat32, parseFtSearchResponse } = await import('../utils');
      const threshold = options?.threshold ?? (this.cache as unknown as { defaultThreshold: number }).defaultThreshold;
      const { vector } = await this.cache._embedText(query);
      const filterExpr = `(@category:{${escapeTag(category)}})`;
      const knnQuery = `${filterExpr}=>[KNN ${limit} @embedding $vec AS __score]`;

      let raw: unknown;
      try {
        raw = await (this.cache as unknown as {
          client: { call: (...args: unknown[]) => Promise<unknown> };
          indexName: string;
        }).client.call(
          'FT.SEARCH',
          (this.cache as unknown as { indexName: string }).indexName,
          knnQuery,
          'PARAMS', '2', 'vec', encodeFloat32(vector),
          'LIMIT', '0', String(limit),
          'DIALECT', '2',
        );
      } catch {
        return [];
      }

      const items: Item[] = [];
      for (const entry of parseFtSearchResponse(raw)) {
        const scoreVal = parseFloat(entry.fields['__score'] ?? 'NaN');
        if (isNaN(scoreVal) || scoreVal > threshold) continue;
        const responseStr = entry.fields['response'];
        if (responseStr) {
          try { items.push(JSON.parse(responseStr) as Item); } catch { /* skip */ }
        }
      }
      return items;
    }

    // No query — return all entries in namespace (up to limit) using _searchEntries
    const { parseFtSearchResponse } = await import('../utils');
    try {
      const result = await this.cache._searchEntries(
        `@category:{${escapeTag(category)}}`, limit, 0,
      );
      const items: Item[] = [];
      for (const entry of parseFtSearchResponse(result)) {
        const responseStr = entry.fields['response'];
        if (responseStr) {
          try { items.push(JSON.parse(responseStr) as Item); } catch { /* skip */ }
        }
      }
      return items;
    } catch {
      return [];
    }
  }

  /**
   * Delete the specific entry at namespace/key.
   * Pages through the namespace using stable SORTBY ordering; adjusts the offset
   * for each deleted entry so no entries are skipped in large namespaces.
   */
  async delete(namespace: string[], key: string): Promise<void> {
    const category = namespaceToCategory(namespace);
    const catFilter = `@category:{${escapeTag(category)}}`;
    const client = (this.cache as unknown as {
      client: { del: (...keys: string[]) => Promise<unknown> };
    }).client;
    const { parseFtSearchResponse } = await import('../utils');
    const BATCH = 100;
    let offset = 0;

    while (true) {
      let raw: unknown;
      try {
        raw = await this.cache._searchEntries(catFilter, BATCH, offset);
      } catch {
        return;
      }
      const entries = parseFtSearchResponse(raw);
      if (entries.length === 0) break;

      let deletedCount = 0;
      for (const entry of entries) {
        const responseStr = entry.fields['response'];
        if (!responseStr) continue;
        try {
          const item = JSON.parse(responseStr) as { key?: string };
          if (item.key === key) {
            await client.del(entry.key).catch(() => { /* best effort */ });
            deletedCount++;
          }
        } catch { /* skip corrupt */ }
      }

      if (entries.length < BATCH) break;
      offset += BATCH - deletedCount;
    }
  }

  /**
   * Batch put/delete multiple items.
   * Executes sequentially to avoid races when the same (namespace, key) appears
   * more than once: concurrent delete+put pairs can interleave and leave duplicates.
   */
  async batch(
    writes: Array<{ namespace: string[]; key: string; value: Record<string, unknown> | null }>,
  ): Promise<void> {
    for (const w of writes) {
      if (w.value === null) {
        await this.delete(w.namespace, w.key);
      } else {
        await this.put(w.namespace, w.key, w.value);
      }
    }
  }
}

File: packages/semantic-cache-py/betterdb_semantic_cache/adapters/langgraph.py
"""LangGraph BaseStore adapter for betterdb-semantic-cache.

BetterDBSemanticStore implements a subset of the LangGraph BaseStore interface,
enabling similarity-based memory retrieval from a SemanticCache instance.

Usage::

    from betterdb_semantic_cache.adapters.langgraph import BetterDBSemanticStore
    store = BetterDBSemanticStore(cache=semantic_cache)
    await store.aput(["user", "alice"], "mem1", {"content": "Alice likes coffee"})
    results = await store.asearch(["user", "alice"], query="What does Alice drink?")
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..semantic_cache import SemanticCache


@dataclass
class Item:
    """A stored item in the semantic memory store."""
    namespace: list[str]
    key: str
    value: dict[str, Any]
    created_at: str
    updated_at: str


def _namespace_key(namespace: list[str]) -> str:
    return ":".join(namespace)


def _namespace_to_category(namespace: list[str]) -> str:
    return _namespace_key(namespace).replace(".", "_").replace("/", "_")


class BetterDBSemanticStore:
    """LangGraph-compatible semantic memory store backed by SemanticCache.

    Args:
        cache: A pre-configured SemanticCache instance.
        embed_field: Field to embed from stored values. Default: 'content'.
    """

    def __init__(self, cache: "SemanticCache", *, embed_field: str = "content") -> None:
        self._cache = cache
        self._embed_field = embed_field

    async def aput(
        self, namespace: list[str], key: str, value: dict[str, Any]
    ) -> None:
        """Store a value at namespace/key."""
        import json
        import time

        embed_text = (
            value[self._embed_field]
            if isinstance(value.get(self._embed_field), str)
            else json.dumps(value)
        )
        category = _namespace_to_category(namespace)
        now = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
        item = Item(namespace=namespace, key=key, value=value,
                    created_at=now, updated_at=now)

        from ..types import CacheStoreOptions
        await self._cache.store(
            embed_text,
            json.dumps({"namespace": namespace, "key": key, "value": value,
                        "created_at": now, "updated_at": now}),
            CacheStoreOptions(
                category=category,
                metadata={"key": key, "namespace": _namespace_key(namespace)},
            ),
        )

    async def aget(self, namespace: list[str], key: str) -> Item | None:
        """Retrieve a value by exact namespace and key."""
        import json

        from ..utils import parse_ft_search_response
        category = _namespace_to_category(namespace)
        try:
            raw = await self._cache._client.execute_command(
                "FT.SEARCH",
                self._cache._index_name,
                f"@category:{{{category}}}",
                "LIMIT", "0", "100",
                "DIALECT", "2",
            )
        except Exception:
            return None

        parsed = parse_ft_search_response(raw)
        for entry in parsed:
            response_str = entry["fields"].get("response")
            if not response_str:
                continue
            try:
                data = json.loads(response_str)
                if data.get("key") == key:
                    return Item(
                        namespace=data.get("namespace", namespace),
                        key=data.get("key", key),
                        value=data.get("value", {}),
                        created_at=data.get("created_at", ""),
                        updated_at=data.get("updated_at", ""),
                    )
            except (json.JSONDecodeError, TypeError):
                pass
        return None

    async def asearch(
        self,
        namespace: list[str],
        *,
        query: str | None = None,
        limit: int = 10,
        threshold: float | None = None,
    ) -> list[Item]:
        """Search the namespace using similarity or scan."""
        import json

        category = _namespace_to_category(namespace)

        if query:
            from ..types import CacheCheckOptions
            results = await self._cache.check_batch(
                [query],
                CacheCheckOptions(category=category, k=limit, threshold=threshold),
            )
            items = []
            for result in results:
                if result.hit and result.response:
                    try:
                        data = json.loads(result.response)
                        items.append(Item(
                            namespace=data.get("namespace", namespace),
                            key=data.get("key", ""),
                            value=data.get("value", {}),
                            created_at=data.get("created_at", ""),
                            updated_at=data.get("updated_at", ""),
                        ))
                    except (json.JSONDecodeError, TypeError):
                        pass
            return items

        # No query — return all entries in namespace
        from ..utils import parse_ft_search_response
        try:
            raw = await self._cache._client.execute_command(
                "FT.SEARCH",
                self._cache._index_name,
                f"@category:{{{category}}}",
                "LIMIT", "0", str(limit),
                "DIALECT", "2",
            )
        except Exception:
            return []

        items = []
        for entry in parse_ft_search_response(raw):
            response_str = entry["fields"].get("response")
            if response_str:
                try:
                    data = json.loads(response_str)
                    items.append(Item(
                        namespace=data.get("namespace", namespace),
                        key=data.get("key", ""),
                        value=data.get("value", {}),
                        created_at=data.get("created_at", ""),
                        updated_at=data.get("updated_at", ""),
                    ))
                except (json.JSONDecodeError, TypeError):
                    pass
        return items

    async def adelete(self, namespace: list[str], key: str) -> None:
        """Delete all entries at namespace/key."""
        await self._cache.invalidate_by_category(_namespace_to_category(namespace))

    async def abatch(
        self,
        writes: list[dict[str, Any]],
    ) -> None:
        """Batch put/delete multiple items."""
        import asyncio

        async def _apply(write: dict[str, Any]) -> None:
            ns = write["namespace"]
            k = write["key"]
            v = write.get("value")
            if v is None:
                await self.adelete(ns, k)
            else:
                await self.aput(ns, k, v)

        await asyncio.gather(*[_apply(w) for w in writes])
